src/lib/controllers/configuration-manager/configuration-manager.ts

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2019 Balena Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as _ from 'lodash';
+
+import { ConfigStoreError, ValidationError } from '../../error-types';
+import { reduce, replace } from 'lodash';
+import { ConfigManifest } from '../config-manifest/config-manifest';
+import { ConfigMap, ConfigStore } from '../config-store/config-store';
+import { validateConfig } from '../config-validator';
+
+interface ConfigurationManagerArgs {
+	configManifest: ConfigManifest;
+	configStore: ConfigStore;
+	configMap: ConfigMap;
+	mode?: string;
+}
+
+interface ConfigurationManagerCreateArgs {
+	configManifest: ConfigManifest;
+	configStore: ConfigStore;
+	configMap?: ConfigMap;
+	mode?: string;
+}
+interface ErrorMap {
+	[key: string]: ValidationError;
+}
+
+/**
+ * ConfigurationManager class
+ * Used for managing configuration stored in a config store, using a config-manifest schema.
+ * Depending on the mode, invalid configuration may be autogenerated, inquired or raise an error.
+ */
+export class ConfigurationManager {
+	/**
+	 * Creates ConfigurationManager using:
+	 * @param {ConfigurationManagerCreateArgs} args
+	 * @returns {Promise<ConfigurationManager>}
+	 */
+	static async create(
+		args: ConfigurationManagerCreateArgs,
+	): Promise<ConfigurationManager> {
+		const {
+			configManifest,
+			configStore,
+			configMap,
+			mode = 'interactive',
+		} = args;
+
+		return new ConfigurationManager({
+			configManifest,
+			configStore,
+			configMap: configMap || (await configStore.list()),
+			mode,
+		});
+	}
+
+	private readonly configManifest: ConfigManifest;
+	private readonly mode: string;
+	private configMap: ConfigMap;
+	private configStore: ConfigStore;
+
+	/**
+	 * ConfigurationManager constructor
+	 * @param {ConfigurationManagerArgs} args
+	 */
+	public constructor(args: ConfigurationManagerArgs) {
+		const { configManifest, configStore, configMap, mode } = args;
+		this.configManifest = configManifest;
+		this.configStore = configStore;
+		this.configMap = configMap;
+		this.mode = mode || 'interactive';
```

Comment:
This can potentially be done via
```ts
const { configManifest, configStore, configMap, mode = 'interactive' } = args;

Reply:
I cannot see a way to do this, but I agree it would be good to find a nicer way to do this kind of pattern.

Reply:
You should be able to do it precisely as I typed if it's acting as a standard default (ie only default when `undefined`) - it'd only be an issue if it's relying on the falsey -> `'interactive'` behaviour

Reply:
ahhh.... @Page- I misunderstood and thought that you were suggesting that the expansion of `args` could directly populate the `this.` fields.

I see now that you were hinting that that default value for `mode` specifically could be done in the expansion, meaning we could drop the `|| 'interactive'` part 👍